libp2p/peeridauth.nim

```diff
@@ -0,0 +1,304 @@
+import base64, json, strutils, uri, times
+import chronos/apps/http/httpclient, results, chronicles, bio
+import ./peerinfo, ./crypto/crypto
+
+logScope:
+  topics = "libp2p peerid auth"
+
+const NimLibp2pUserAgent = "nim-libp2p"
+const PeerIDAuthPrefix = "libp2p-PeerID"
+const ChallengeCharset =
+  "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+const ChallengeDefaultLen = 48
+
+type PeerIDAuthError* = object of LPError
+
+type BearerToken* = object
+  token*: string
+  expires*: Opt[DateTime]
+
+type PeerIDAuthChallenge = string
+type PeerIDAuthOpaque = string
+type PeerIDAuthSignature = string
+
+type PeerIDAuthApi* = object
+  session: HttpSessionRef
+  rng: ref HmacDrbgContext
+
+type PeerIDAuthAuthenticationResponse* = object
+  challengeClient: PeerIDAuthChallenge
+  opaque: PeerIDAuthOpaque
+  serverPubkey: PublicKey
+
+type PeerIDAuthAuthorizationResponse* = object
+  sig: PeerIDAuthSignature
+  bearer: BearerToken
+  rawResponse: HttpClientResponseRef
+
+type SigParam = object
+  k: string
+  v: seq[byte]
+
+proc new*(T: typedesc[PeerIDAuthApi], rng: ref HmacDrbgContext): PeerIDAuthApi =
+  PeerIDAuthApi(session: HttpSessionRef.new(), rng: rng)
+
+proc sampleChar(ctx: var HmacDrbgContext, choices: string): char =
+  ## Samples a random character from the input string using the DRBG context
+  if choices.len == 0:
+    raise newException(ValueError, "Cannot sample from an empty string")
+  var idx: uint32
+  ctx.generate(idx)
+  return choices[uint32(idx mod uint32(choices.len))]
+
+proc randomChallenge(
+    rng: ref HmacDrbgContext, challengeLen: int = ChallengeDefaultLen
+): PeerIDAuthChallenge {.raises: [PeerIDAuthError].} =
+  var rng = rng[]
+  var challenge = ""
+  try:
+    for _ in 0 ..< challengeLen:
+      challenge.add(rng.sampleChar(ChallengeCharset))
+  except ValueError as exc:
+    raise newException(PeerIDAuthError, "Failed to generate challenge", exc)
+  PeerIDAuthChallenge(challenge)
+
+proc extractField(data, key: string): string {.raises: [PeerIDAuthError].} =
+  # Helper to extract quoted value from key
+  for segment in data.split(","):
+    if key in segment:
+      return segment.split("=", 1)[1].strip(chars = {' ', '"'})
+  raise newException(PeerIDAuthError, "Failed to find " & key & " in " & data)
+
+proc encodeVarint(n: int): seq[byte] =
+  var varInt: seq[byte] = @[]
+  var x = uint64(n)
+  while x != 0:
+    var byteVal = byte(x and 0x7F)
+    x = x shr 7
+    if x != 0:
+      byteVal = byteVal or 0x80
+    varInt.add(byteVal)
+  return varInt
+
+proc genDataToSign(parts: seq[SigParam], prefix: string = PeerIDAuthPrefix): seq[byte] =
+  var buf: seq[byte] = prefix.toByteSeq()
+  for p in parts:
+    buf.add encodeVarint(p.k.len + p.v.len + 1)
+    buf.add (p.k & "=").toByteSeq()
+    buf.add p.v
```

Comment:
Couldnt this and encodeVarint be replaced by libp2p varint.nim?
with something like 
```nim
proc genDataToSign(parts: seq[SigParam], prefix: string = PeerIDAuthPrefix): seq[byte] =
  var buf: seq[byte] = prefix.toByteSeq()
  for p in parts:
    buf.add LP.encodeVarint(p.k.len + p.v.len + 1)
    buf.add (p.k & "=").toByteSeq()
    buf.add p.v
```